

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>metapredict.meta &mdash; metapredict  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> metapredict
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started with metapredict</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/graph-disorder.html">graph-disorder</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/predict-disorder.html">predict-disorder</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/using-in-python.html">metapredict in Python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started with metapredict</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">Module Documentation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">metapredict</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>metapredict.meta</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for metapredict.meta</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">meta.py</span>
<span class="sd">A protein disorder predictor based on a BRNN (IDP-Parrot) trained</span>
<span class="sd">on the consensus disorder values from 8 disorder predictors from </span>
<span class="sd">12 proteomes (see https://mobidb.bio.unipd.it) as of 08/2020.</span>

<span class="sd">Handles the primary functions</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="c1">#import protfasta to read .fasta files</span>
<span class="kn">import</span> <span class="nn">protfasta</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1">#import stuff for IDR predictor from backend</span>
<span class="kn">from</span> <span class="nn">metapredict.backend</span> <span class="kn">import</span> <span class="n">meta_predict_disorder</span>
<span class="kn">from</span> <span class="nn">metapredict.backend.meta_predict_disorder</span> <span class="kn">import</span> <span class="n">meta_predict</span>

<span class="c1">#import stuff for graphing from backend</span>
<span class="kn">from</span> <span class="nn">metapredict.backend</span> <span class="kn">import</span> <span class="n">meta_graph</span>
<span class="kn">from</span> <span class="nn">metapredict.backend.meta_graph</span> <span class="kn">import</span> <span class="n">graph</span>



<div class="viewcode-block" id="predict_disorder"><a class="viewcode-back" href="../../api.html#metapredict.meta.predict_disorder">[docs]</a><span class="k">def</span> <span class="nf">predict_disorder</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to return disorder of a single input sequence. Returns the</span>
<span class="sd">    predicted values as a list.</span>

<span class="sd">    Arguments:</span>
<span class="sd">    ----------</span>
<span class="sd">    sequence - Input amino acid sequence (as string) to be predicted.</span>

<span class="sd">    normalized (optional) - by default predictor returns values normalized between 0 and 1.</span>
<span class="sd">    This is because the BRNN will output some negative values and some values greater</span>
<span class="sd">    than 1. Setting normalized=False will result in returning the raw predicted values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#make all residues upper case </span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="c1">#return predicted values of disorder for sequence</span>
    <span class="k">return</span> <span class="n">meta_predict</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="n">normalized</span><span class="p">)</span></div>


<div class="viewcode-block" id="graph_disorder"><a class="viewcode-back" href="../../api.html#metapredict.meta.graph_disorder">[docs]</a><span class="k">def</span> <span class="nf">graph_disorder</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">DPI</span><span class="o">=</span><span class="mi">150</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to plot the disorder of an input sequece. Displays immediately.</span>

<span class="sd">    Arguments:</span>
<span class="sd">    ----------</span>
<span class="sd">    sequence - Input amino acid sequence (as string) to be predicted.</span>

<span class="sd">    name (optional) - setting the value of name will change the title of the</span>
<span class="sd">    graph. By default, the title is &quot;Predicted Protein Disorder&quot;, so if you</span>
<span class="sd">    for example set name = &quot;- PAB1&quot;, the title on the graph will be &quot;Predicted</span>
<span class="sd">    Protein Disorder - PAB1&quot;. </span>

<span class="sd">    DPI (optional) - default value is 150. Increasing this value will increase</span>
<span class="sd">    the resolution of the output graph. Decreasing this value will decrease</span>
<span class="sd">    the resolution.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#make all residues upper case </span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="c1">#graph sequence</span>
    <span class="n">graph</span><span class="p">(</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="n">DPI</span><span class="o">=</span><span class="n">DPI</span><span class="p">)</span></div>



<div class="viewcode-block" id="percent_disorder"><a class="viewcode-back" href="../../api.html#metapredict.meta.percent_disorder">[docs]</a><span class="k">def</span> <span class="nf">percent_disorder</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.5</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    function to return the percent disorder for any given protein.</span>
<span class="sd">    By default, uses 0.5 as a cutoff (values greater than or equal</span>
<span class="sd">    to 0.5 will be considred disordered).</span>
<span class="sd">    </span>
<span class="sd">    Arguments:</span>
<span class="sd">    ----------</span>
<span class="sd">    sequence - Input amino acid sequence (as string) to be predicted.</span>
<span class="sd">    </span>
<span class="sd">    cutoff (optional) the cutoff for the predicted value of an individual</span>
<span class="sd">    residue to be considered disordered. By default this value is 0.5. Increasing</span>
<span class="sd">    this value will make the cutoff more &quot;strict&quot; in that a higher predicted</span>
<span class="sd">    vallue will be required for a residue to be considered disordered.</span>

<span class="sd">    Returns the percent disorder for the input sequence as a decimal. </span>
<span class="sd">    1.0 = 100% disordered,</span>
<span class="sd">    0.9 = 90% disordered, </span>
<span class="sd">    and so on.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#make all residues upper case </span>
    <span class="n">sequence</span><span class="o">=</span><span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="c1">#set dis equal to the predicted disorder for the input sequence</span>
    <span class="n">dis</span> <span class="o">=</span> <span class="n">meta_predict</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="c1">#set arbitrarily chosen variable n to equal 0</span>
    <span class="n">n</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1">#for predicted disorder values in dis:</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">dis</span><span class="p">:</span>
        <span class="c1">#if predicted value is greater than cutoff, add one to n</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="n">cutoff</span><span class="p">:</span>
            <span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="c1">#else continue through the values.</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">continue</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    percent disorder is equal to n (number of residues with predicted</span>
<span class="sd">    value &gt;= cutoff) divided by the total number of residues in the</span>
<span class="sd">    input sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">percent_disordered</span> <span class="o">=</span> <span class="p">(</span><span class="n">n</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">dis</span><span class="p">))</span>
    <span class="c1">#return percent_disordered</span>
    <span class="k">return</span><span class="p">(</span><span class="n">percent_disordered</span><span class="p">)</span></div>



<span class="c1">#./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\</span>
<span class="c1">#./\./\./\./\./\./\./\./\./\./\./\./\.FASTA STUFF./\./\./\./\./\./\./\./\./\./\./\./\</span>
<span class="c1">#./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\./\</span>

<span class="c1">#Various functions for working with fasta files to make everyones life easier.</span>


<div class="viewcode-block" id="predict_disorder_fasta"><a class="viewcode-back" href="../../api.html#metapredict.meta.predict_disorder_fasta">[docs]</a><span class="k">def</span> <span class="nf">predict_disorder_fasta</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">output_path</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">output_name</span> <span class="o">=</span> <span class="s2">&quot;predicted_disorder_values&quot;</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to read in a .fasta file from a specified filepath.</span>
<span class="sd">    Returns a dictionary of disorder values where the key is the </span>
<span class="sd">    fasta header and the values are the predicted disorder values.</span>
<span class="sd">    </span>
<span class="sd">    Arguments:</span>
<span class="sd">    ----------</span>
<span class="sd">    filepath - the path to where the .fasta file is located. The filepath</span>
<span class="sd">    should end in the file name. For example (on MacOS):</span>
<span class="sd">    filepath=&quot;/Users/thisUser/Desktop/folder_of_seqs/interesting_proteins.fasta&quot;</span>
<span class="sd">    </span>
<span class="sd">    save (optional) - by default, a dictionary of predicted values is </span>
<span class="sd">    returned immediately. However, you can specify save=True in order to </span>
<span class="sd">    save the output as a .csv file. </span>
<span class="sd">    ***important***</span>
<span class="sd">    If you specify save=True, then output_path is a required argument!</span>

<span class="sd">    output_path - the path to where the output .csv file should be saved. </span>
<span class="sd">    For example, on MacOS:</span>
<span class="sd">    output_path=&quot;Users/thisUser/Desktop/folder_of_cool_results/&quot;</span>
<span class="sd">    ***important***</span>
<span class="sd">    You cannot specify the output file name here! By default, the file name will</span>
<span class="sd">    be predicted_values.csv. However, you can change the output file name by </span>
<span class="sd">    specifying ouput_name (see below).</span>
<span class="sd">    </span>
<span class="sd">    output_name (optional) - by default is set to equal &quot;predicted values&quot;. However,</span>
<span class="sd">    the user can specify output_name=&quot;my_cool_output_name&quot; in order to specify the</span>
<span class="sd">    name of the output file.</span>
<span class="sd">    ***important***</span>
<span class="sd">    Do not add a file extension to output_name. The .csv file extension is added </span>
<span class="sd">    automatically. If you do by accident, it will (hopefully) be removed.</span>

<span class="sd">    normalized - decide whether the values are normalized from 0 to 1. By default, the</span>
<span class="sd">    values are normalized. However, occassionally the raw values from the predictor can be</span>
<span class="sd">    negative or greater than 1. If you want thes values, set normalied=False.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#set variable protfastaSeqs equal to output from protfasta (with correction of invalid sequence values)</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Importantly, by default this function corrects invalid residue</span>
<span class="sd">    values using protfasta.read_fasta() because the disorder predictor</span>
<span class="sd">    cannot have non-amino acid values as an input.</span>
<span class="sd">    &quot;&quot;&quot;</span>    
    <span class="c1"># Test to see if the data_file exists</span>
    <span class="n">test_data_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">test_data_file</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s1">&#39;Datafile does not exist.&#39;</span><span class="p">)</span>

    <span class="n">protfasta_seqs</span> <span class="o">=</span> <span class="n">protfasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">invalid_sequence_action</span> <span class="o">=</span> <span class="s2">&quot;convert&quot;</span><span class="p">,</span> <span class="n">return_list</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="c1">#initialize empty dictionary to be populated with the the fasta headers (key) </span>
    <span class="c1">#and the predicted disorder values (value)</span>
    <span class="n">disorder_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1">#for the sequences in the protffasta_seqs list:</span>
    <span class="k">for</span> <span class="n">seqs</span> <span class="ow">in</span> <span class="n">protfasta_seqs</span><span class="p">:</span>
        <span class="c1">#set cur_header equal to the fasta header</span>
        <span class="n">cur_header</span> <span class="o">=</span> <span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="c1">#set cur_seq equal to the sequence associated with the fasta header</span>
        <span class="n">cur_seq</span> <span class="o">=</span> <span class="n">seqs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="c1">#make all values for curSeq uppercase so they work with predictor</span>
        <span class="n">cur_seq</span> <span class="o">=</span> <span class="n">cur_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="c1">#set cur_disorder equal to the predicted values for cur_seq</span>
        <span class="n">cur_disorder</span> <span class="o">=</span> <span class="n">meta_predict</span><span class="p">(</span><span class="n">cur_seq</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="n">normalized</span><span class="p">)</span>
        <span class="n">disorder_dict</span><span class="p">[</span><span class="n">cur_header</span><span class="p">]</span> <span class="o">=</span> <span class="n">cur_disorder</span>

    <span class="c1">#if save=False (default), immediately return the dictionary disorder_dict</span>
    <span class="k">if</span> <span class="n">save</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">disorder_dict</span>

    <span class="c1">#if save=True, save the disorder_dict to the specified output_path</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Test to see that the output path is valid</span>
        <span class="n">test_output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">test_output_path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s1">&#39;Output path is not valid.&#39;</span><span class="p">)</span>

        <span class="c1">#Check if there is a .csv in output_name (which there shouldn&#39;t be)</span>
        <span class="c1">#set try_output_name = output_name</span>
        <span class="n">try_output_name</span> <span class="o">=</span> <span class="n">output_name</span>
        <span class="c1">#if there is .csv in try_output_name</span>
        <span class="k">if</span> <span class="s2">&quot;.csv&quot;</span> <span class="ow">in</span> <span class="n">try_output_name</span><span class="p">:</span>
            <span class="c1">#split try_output_name and set output_file_name equal to everything before .csv</span>
            <span class="n">output_file_name</span> <span class="o">=</span> <span class="n">try_output_name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.csv&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#if .csv is not in try_output_name, set final output_final_name equal to args.output_name</span>
            <span class="n">output_file_name</span> <span class="o">=</span> <span class="n">output_name</span>

        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Make sure output_path ends in / (mac) or \\ (windows).</span>
<span class="sd">        This is necessary because earlier when testing the output path using OS,</span>
<span class="sd">        a valid output path can still not end in a / or \\. When the path does not</span>
<span class="sd">        end in a / or \\, then the file does not get saved correctly but pandas</span>
<span class="sd">        doesn&#39;t raise an error. I&#39;m not sure this is totally necessary, but</span>
<span class="sd">        I made this mistake a few times while testing this stuff so I figured it might</span>
<span class="sd">        be a nice feature.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">#Set final_output_path_character = last character in output_path</span>
        <span class="n">final_output_path_character</span> <span class="o">=</span> <span class="n">output_path</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="c1">#if / in output_path (user is using MacOS or linux)</span>
        <span class="k">if</span> <span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="n">output_path</span><span class="p">:</span>
            <span class="c1">#if the final character in the path does not equal a /</span>
            <span class="k">if</span> <span class="n">final_output_path_character</span> <span class="o">!=</span> <span class="s2">&quot;/&quot;</span><span class="p">:</span>
                <span class="c1">#add in a / to complete file path</span>
                <span class="n">output_path</span> <span class="o">+=</span> <span class="s2">&quot;/&quot;</span>
        <span class="c1">#if \ in output_path (user is using Windows)</span>
        <span class="k">elif</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">&quot;</span> <span class="ow">in</span> <span class="n">output_path</span><span class="p">:</span>
            <span class="c1">#if the last character does not equal \</span>
            <span class="k">if</span> <span class="n">final_output_path_character</span> <span class="o">!=</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">&quot;</span><span class="p">:</span>
                <span class="c1">#add in a \ to output path</span>
                <span class="n">output_path</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">&quot;</span>

        <span class="c1">#make pandas dataframe from disorder_dict</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">disorder_dict</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">)</span>
        <span class="c1">#set variable final_output equal to the output_path followed by output_name</span>
        <span class="c1">#(by default output_name is equal to predicted_disorder_values).</span>
        <span class="n">final_output</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}{}</span><span class="s2">.csv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">output_file_name</span><span class="p">)</span>
        <span class="c1">#export the dataframe to a csv file at the location final_output</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">final_output</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></div>



<div class="viewcode-block" id="graph_disorder_fasta"><a class="viewcode-back" href="../../api.html#metapredict.meta.graph_disorder_fasta">[docs]</a><span class="k">def</span> <span class="nf">graph_disorder_fasta</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">DPI</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">output_path</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">remove_characters</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to make graphs of predicted disorder from the sequences</span>
<span class="sd">    in a specified .fasta file. By default will save the generated</span>
<span class="sd">    graphs to the location output_path specified in filepath.</span>
<span class="sd">    </span>
<span class="sd">    Arguments:</span>
<span class="sd">    ----------</span>
<span class="sd">    filepath - the path to where the .fasta file is located. The filepath</span>
<span class="sd">    should end in the file name. For example (on MacOS):</span>
<span class="sd">    filepath=&quot;/Users/thisUser/Desktop/folder_of_seqs/interesting_proteins.fasta&quot;</span>
<span class="sd">    </span>
<span class="sd">    DPI (optional) - default value is 150. Increasing this value will increase</span>
<span class="sd">    the resolution of the output graph. Decreasing this value will decrease</span>
<span class="sd">    the resolution. </span>

<span class="sd">    save (optional) - by default, the generated graphs are saved. This can be set</span>
<span class="sd">    to False, which will result in the graphs being sequentially shown.</span>
<span class="sd">    ***important***</span>
<span class="sd">    It is unadvisable to set save=False if you are inputting a large .fasta file! This</span>
<span class="sd">    is because each graph must be closed individually before the next will appear. Therefore,</span>
<span class="sd">    you will spend a bunch of time closing each graph.</span>

<span class="sd">    output_path - the path to where the output graphs should be saved. </span>
<span class="sd">    For example, on MacOS:</span>
<span class="sd">    output_path=&quot;Users/thisUser/Desktop/folder_of_cool_results/&quot;</span>
<span class="sd">    ***important***</span>
<span class="sd">    You cannot specify the output file name here! By default, the file name will</span>
<span class="sd">    be predicted_disorder followed by 6 characters from the .fasta header (which ideally</span>
<span class="sd">    should be a unique identifier that can be input into uniprot) followed by .png. </span>
<span class="sd">    For example, predicted_disorder_sp|O43150|.png</span>

<span class="sd">    remove_characters (optional) - allows all non-alphabetic characters to be removed</span>
<span class="sd">    from the output file names. If user is on an OS that doesn&#39;t allow specific characters</span>
<span class="sd">    to be used in file names, this should be a way to get around that.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Test to see if the data_file exists</span>
    <span class="n">test_data_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">test_data_file</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s1">&#39;Datafile does not exist.&#39;</span><span class="p">)</span>

    <span class="c1">#use protfasta to read in fasta file</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Importantly, by default this function corrects invalid residue</span>
<span class="sd">    values using protfasta.read_fasta() because the disorder predictor</span>
<span class="sd">    cannot have non-amino acid values as an input.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sequences</span> <span class="o">=</span> <span class="n">protfastaSeqs</span> <span class="o">=</span> <span class="n">protfasta</span><span class="o">.</span><span class="n">read_fasta</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">invalid_sequence_action</span> <span class="o">=</span> <span class="s2">&quot;convert&quot;</span><span class="p">)</span>

    <span class="c1">#for key, value in sequences.items (which are the items in the dict returned by protfasta)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">sequences</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="c1">#set title of graph equal to the first 10 characters of the fasta header after &gt;</span>
        <span class="n">title</span> <span class="o">=</span> <span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>
        <span class="c1">#if remove_characters is False:</span>
        <span class="k">if</span> <span class="n">remove_characters</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
            <span class="c1">#set name equal to the first 10 characters of the fasta header after &gt;</span>
            <span class="n">name</span> <span class="o">=</span> <span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#initializing empty string</span>
            <span class="n">empty_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="c1">#removing all common characters (non-alphabetic) from the fasta header</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">j</span><span class="o">==</span><span class="s2">&quot;&gt;&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">elif</span> <span class="n">j</span><span class="o">==</span><span class="s2">&quot;|&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">elif</span> <span class="n">j</span><span class="o">==</span><span class="s2">&quot;=&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">elif</span> <span class="n">j</span><span class="o">==</span><span class="s2">&quot;-&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">elif</span> <span class="n">j</span><span class="o">==</span><span class="s2">&quot; &quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">empty_name</span> <span class="o">+=</span> <span class="n">j</span>
            <span class="c1">#set final name equal to the first 10 characters from the fasta header with</span>
            <span class="c1">#the various characters removed.</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">empty_name</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>

        <span class="c1">#set the sequence equal to the amino acid sequence associated with</span>
        <span class="c1">#the fasta header and make all amino acids uppercase.</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="c1">#If save is True (default)</span>
        <span class="k">if</span> <span class="n">save</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="c1"># Test to see that the output path is valid</span>
            <span class="n">test_output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">test_output_path</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="s1">&#39;Output path is not valid.&#39;</span><span class="p">)</span>


            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Make sure output_path ends in / (mac) or \\ (windows).</span>
<span class="sd">            This is necessary because earlier when testing the output path using OS,</span>
<span class="sd">            a valid output path can still not end in a / or \\. When the path does not</span>
<span class="sd">            end in a / or \\, then the file does not get saved correctly but pandas</span>
<span class="sd">            doesn&#39;t raise an error. I&#39;m not sure this is totally necessary, but</span>
<span class="sd">            I made this mistake a few times while testing this stuff so I figured it might</span>
<span class="sd">            be a nice feature.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="c1">#Set final_output_path_character = last character in output_path</span>
            <span class="n">final_output_path_character</span> <span class="o">=</span> <span class="n">output_path</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="c1">#if / in output_path (user is using MacOS or linux)</span>
            <span class="k">if</span> <span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="n">output_path</span><span class="p">:</span>
                <span class="c1">#if the final character in the path does not equal a /</span>
                <span class="k">if</span> <span class="n">final_output_path_character</span> <span class="o">!=</span> <span class="s2">&quot;/&quot;</span><span class="p">:</span>
                    <span class="c1">#add in a / to complete file path</span>
                    <span class="n">output_path</span> <span class="o">+=</span> <span class="s2">&quot;/&quot;</span>
            <span class="c1">#if \ in output_path (user is using Windows)</span>
            <span class="k">elif</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">&quot;</span> <span class="ow">in</span> <span class="n">output_path</span><span class="p">:</span>
                <span class="c1">#if the last character does not equal \</span>
                <span class="k">if</span> <span class="n">final_output_path_character</span> <span class="o">!=</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">&quot;</span><span class="p">:</span>
                    <span class="c1">#add in a \ to output path</span>
                    <span class="n">output_path</span> <span class="o">+=</span> <span class="s2">&quot;</span><span class="se">\\</span><span class="s2">&quot;</span>

            <span class="c1">#set variable output equal to the output_path folowed by the file name which is</span>
            <span class="c1">#predicted_disorder_{name}.png, where name is the name specified earlier (first 10 characters</span>
            <span class="c1">#of the fasta header either as is or with characters removed if remove_characters is set to true).</span>
            <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">predicted_disorder_</span><span class="si">{}</span><span class="s2">.png&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
            <span class="c1">#use the graph function (specified in meta_graph from the backend) to save the graph.</span>
            <span class="n">graph</span><span class="p">(</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="n">title</span><span class="p">,</span> <span class="n">DPI</span> <span class="o">=</span> <span class="n">DPI</span><span class="p">,</span> <span class="n">save_fig</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">output_file</span> <span class="o">=</span> <span class="n">output</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#if save was set to False, then just graph the sequences from the .fasta file and show them immediately.</span>
            <span class="n">graph</span><span class="p">(</span><span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="n">title</span><span class="p">,</span> <span class="n">DPI</span> <span class="o">=</span> <span class="n">DPI</span><span class="p">)</span></div>

</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Ryan Emenecker - Holehouse Lab - WUSM. Project structure based on the Computational Molecular Science Python Cookiecutter version 1.3

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>